components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,515 @@
+                                    resultt->request_id)) {
+        return {};
+      }
+    }
+    if (resultt->request_id &&
+        !base::ranges::all_of(*resultt->request_id, &IsPChar)) {
+      return {};
+    }
+    if (index != message.size()) {
+      next_segment = GetNextSegment(message, index);
+      if (!next_segment ||
+          !ParseOptionalResources(*next_segment, index, resultt->resources)) {
+        return {};
+      }
+    }
+  }
+  if (index != message.size()) {
+    return {};
+  }
+  state_ = State::kEnd;
+  return resultt;
+}
+
+bool SIWEMessageParser::ParseSchemeAndDomain(std::string_view msg_view,
+                                             size_t& index,
+                                             url::Origin& origin) {
+  std::string::size_type n =
+      msg_view.find(base::StrCat({kStartingToken, "\n"}));
+  if (!n || n == std::string::npos) {
+    return false;
+  }
+  std::string_view old_origin_str(msg_view.begin(), msg_view.begin() + n);
+  std::string new_origin_str(old_origin_str);
+  // If scheme is not specified, https will be used by default.
+  if (!base::Contains(old_origin_str, url::kStandardSchemeSeparator)) {
+    new_origin_str = base::StrCat(
+        {url::kHttpsScheme, url::kStandardSchemeSeparator, old_origin_str});
+  }
+  GURL url(new_origin_str);
+  if (!url.is_valid()) {
+    return false;
+  }
+
+  origin = url::Origin::Create(url);
+  index += (n + strlen(kStartingToken) + 1);
+  state_ = State::kAddress;
+  return true;
+}
+
+bool SIWEMessageParser::ParseAddress(std::string_view msg_view,
+                                     size_t& index,
+                                     std::string& address_out) {
+  std::string::size_type n = msg_view.find('\n');
+  if (!n || n == std::string::npos) {
+    return false;
+  }
+  const std::string address(msg_view.begin(), msg_view.begin() + n);
+  if (!EthAddress::IsValidAddress(address)) {
+    return false;
+  }
+  index += (n + 1);
+  address_out = address;
+  state_ = State::kStatement;
+  return true;
+}
+
+bool SIWEMessageParser::ParseStatement(std::string_view msg_view,
+                                       size_t& index,
+                                       absl::optional<std::string>& statement) {
+  if (msg_view.size() < 2 || msg_view[0] != '\n') {
+    return false;
+  }
+  if (msg_view[1] == '\n') {
+    // jump forward "\n\n"
+    index += 2;
+    state_ = State::kURI;
+    // no statement
+    return true;
+  }
+  std::string::size_type stmt_lf = msg_view.find('\n', 1);
+  if (stmt_lf == std::string::npos) {
+    return false;
+  }
+  // closing lf
+  if (stmt_lf + 1 >= msg_view.length() || msg_view[stmt_lf + 1] != '\n') {
+    return false;
+  }
+  statement = std::string(msg_view.begin() + 1, msg_view.begin() + stmt_lf);
+  if (!base::IsStringASCII(*statement)) {
+    return false;
+  }
+  index += (stmt_lf + 2);
+  state_ = State::kURI;
+  return true;
+}
```

Comment:
In case we have parse by line logic like:
```
std::string_view ExtractNextLine(std::string_view& from) {
  auto pos = from.find('\n');
  if (pos == std::string_view::npos) {
    auto result = from;
    from.clear();
    return result;
  } 
  auto result = from.substr(0, pos); // trim lf, we don't need as it is part of grammar
  from.remove_prefix(pos + 1);
  return result;
}
```
(Note: that's ok to read to the end in case there is no '\n'. It that case the next REQUIRED line will be empty and fail whole parsing)

We can parse it simpler:
```
bool SIWEMessageParser::ParseStatement(std::string_view& msg_view, absl::optional<std::string>& statement){
  auto line = ExtractNextLine(msg_view); // first empty line
  if (!line.empty()){
    return false;
  }
  line = ExtractNextLine(msg_view); // possible second empty line.
  if (line.empty()) {
    return true;
  }

  statement = line; // not empty, so it is statement. (needs ascii check)
  if (!ExtractNextLine(msg_view).empty()){ // second empty line.
    return false;
  }
}
```

Reply:
This `ExtractNextLine` approach only works for require fields and optional statement because they are all ended with `\n` but not other optional fields which are all starting with `\n` and also last required field, `IssuedAt`. 
I don't see it would be simpler than letting each parsing part extract and consume what they need and leave rest of the unhandled message for the next segment.